app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +28,165 @@ pub use example::Example;
+/// formed of segments with no guaranteed semantic meaning.
+///
+/// A conversion from a string value is performed by splitting the string into segments separated
+/// by the [`ACCESS`] character.
+#[derive(Clone, Debug, Default)]
+pub struct FreeformPath {
+    segments: Vec<PathSegment>,
+}
+
+impl FreeformPath {
+    fn from_segments<I, S>(segments: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<PathSegment>, {
+        let segments = segments.into_iter().map(|s| s.into()).collect();
+        Self { segments }
+    }
+}
+
+impl From<&str> for FreeformPath {
+    fn from(path: &str) -> FreeformPath {
+        FreeformPath::from_segments(path.split(ACCESS))
+    }
+}
+
+impl From<String> for FreeformPath {
+    fn from(path: String) -> FreeformPath {
+        path.as_str().into()
+    }
+}
+
+impl From<&Entry> for FreeformPath {
+    fn from(entry: &Entry) -> FreeformPath {
+        match entry.kind {
+            Kind::Method => {
+                let mut path = match &entry.self_type {
+                    Some(name) => FreeformPath::from_segments(name.segments()),
+                    None => default(),
+                };
+                path.segments.push(entry.name.clone().into());
+                path
+            }
+            Kind::Module => FreeformPath::from_segments(entry.module.segments()),
+            _ => {
+                let mut path = FreeformPath::from_segments(entry.module.segments());
+                path.segments.push(entry.name.clone().into());
+                path
+            }
+        }
+    }
+}
+
+
+
+// ===========================
+// === FreeformPathToIdMap ===
+// ===========================
+
+#[derive(Clone, Debug, Default)]
+struct FreeformPathToIdMap {
+    tree: RefCell<ensogl::data::HashMapTree<PathSegment, Option<entry::Id>>>,
+}
+
+impl FreeformPathToIdMap {
+    fn warn_if_exists_and_set(
+        &self,
+        path: impl Into<FreeformPath>,
+        id: entry::Id,
+        logger: &Logger,
+    ) {
+        let path = path.into();
+        let existed = self.check_if_exists_and_set(&path, id);
+        if existed {
+            let path = path.segments;
+            warning!(logger, "An existing id at {path:?} was overwritten with {id}.");
+        }
+    }
+
+    fn warn_if_absent_and_remove(&self, path: impl Into<FreeformPath>, logger: &Logger) {
+        let path = path.into();
+        let existed = self.check_if_exists_and_remove(&path);
+        if !existed {
+            let path = path.segments;
+            let msg =
+                format!("When removing an id at {path:?} some id was expected but none was found.");
+            warning!(logger, "{msg}");
+        }
+    }
+
+    fn check_if_exists_and_set(&self, path: &FreeformPath, id: entry::Id) -> bool {
+        let mut tree = self.tree.borrow_mut();
+        let node = &mut tree.get_or_create_node(&path.segments);
+        let old_value = std::mem::replace(&mut node.value, Some(id));
```

Comment:
all of these `check_if_...` and `warn_if_...` functions are repetitive. Think about refactoring them to a common base function.